ext/hash/hash_murmur.c

```diff
@@ -82,9 +93,24 @@ const php_hash_ops php_hash_murmur3c_ops = {
 	0
 };
 
-PHP_HASH_API void PHP_MURMUR3CInit(PHP_MURMUR3C_CTX *ctx)
+PHP_HASH_API void PHP_MURMUR3CInit(PHP_MURMUR3C_CTX *ctx, HashTable *args)
 {
-	memset(&ctx->h, 0, sizeof ctx->h);
+	if (args) {
+		zval *seed = zend_hash_str_find_deref(args, "seed", sizeof("seed") - 1);
+		/* This might be a bit too restrictive, but thinking that a seed might be set
+			once and for all, it should be done a clean way. */
+		if (seed && IS_LONG == Z_TYPE_P(seed)) {
+			uint32_t _seed = (uint32_t)Z_LVAL_P(seed);
+			ctx->h[0] = _seed;
+			ctx->h[1] = _seed;
+			ctx->h[2] = _seed;
+			ctx->h[3] = _seed;
```

Comment:
One more question: Is this seeding approach "standard" in some way, or just chosen arbitrarily?

E.g. it would be possible to accept a 16 byte string here and seed the whole state from that, without repeating a value. But I have no idea whether that makes more or less sense than what you're doing right now.

Reply:
Yep, this mimics the canonical implementation. Observe here

https://github.com/aappleby/smhasher/blob/master/src/MurmurHash3.cpp#L150
https://github.com/aappleby/smhasher/blob/master/src/MurmurHash3.cpp#L255

The difference in the implementation currently in PHP is that it separates init/update/result steps, so then it can save memory and support streaming. The produced hashes are compatible with anything based on the canonical implementation.

Technically, the approach of passing different values for each of the initial state parts could be considered later as a non standard extension to this. Though, the murmur implementation does a good job here by making a dependency on the data. Say in the case 128-bit for x86 there are 4 32-bit parts. If the data has not an appropriately aligned length, the algorithm will modify those 32-bit parts depending on the trailing data part.

Thanks.

Reply:
Thanks for the explanation!